lib/services/lookup.js

```diff
@@ -96,6 +99,115 @@ function lookupServiceFactory(
+            this.macRequests = {};
+        } else {
+            delete this.macRequests;
+        }
+    };
+
+    /**
+     * Look up an entry in the lookups collection by IP. Optionally fall back
+     * to other mechanisms if we can't find an entry.
+     * @private
+     * @param {ip} IP address to look for
+     * @returns lookups record
+     */
+    LookupService.prototype.lookupByIP = function (ip) {
+        return this.validateArpCache()
+        .then(function() {
+            return waterline.lookups.findOneByTerm(ip);
+        });
+    };
+
+    /**
+     * Look up an entry in the lookups collection by MAC. Optionally fall back
+     * to other mechanisms if we can't find an entry.
+     * @private
+     * @param {mac} MAC address to look for
+     * @returns lookups record
+     */
+    LookupService.prototype.lookupByMAC = function (mac) {
+        return this.validateArpCache()
+        .then(function() {
+            return waterline.lookups.findOneByTerm(mac);
+        })
+        .catch(function (error) {
+            var helper = configuration.get('externalLookupHelper', null);
+            if (helper != null) {
+                return this.runExternalHelper(helper, mac).then(function () {
+                    return waterline.lookups.findOneByTerm(mac);
+                });
+            }
+            throw error;
+        }.bind(this));
+    };
+
+    /**
+     * Run an external helper script to fill in missing lookup entries.
+     * Only one instance of the helper script is run at any given time for
+     * any given MAC address. All other requests will be resolved when the
+     * running instance returns.
+     * @private
+     * @param {helperPath} Path to the lookup script.
+     * @param {mac} The MAC address we need information on.
+     */
+    LookupService.prototype.runExternalHelper = function (helperPath, mac) {
+        if (mac in this.macRequests) {
+            return new Promise(function (resolve, reject) {
+                var responder = {resolve: resolve, reject: reject};
+                this.macRequests[mac].push(responder);
+            }.bind(this));
+        }
+
+        var helper = new ChildProcess(helperPath, [mac]);
+        this.macRequests[mac] = [];
+        return this.processHelperResults(helper).then(function () {
+            this.handleHelperPromises(mac)
+        }.bind(this));
+    };
+
+    /**
+     * Process the output from an external lookup helper script.
+     * The expected output is one or more lines of text of the form:
+     *    <MAC address> <IP address>
+     * Each entry will be turned into arguments for this.setIpAddress.
+     * @private
+     * @param {helper} The ChildProcess object representing the helper script.
+     */
+    LookupService.prototype.processHelperResults = function (helper) {
+        var self = this;
+        return helper.run().then(function (ret) {
+            var promises = [];
+
+            ret.stdout.split('\n').forEach(function (line) {
```

Comment:
@jeamland Just as an FYI - you can use Promise.map in place of a promise array / Promise.all.  http://bluebirdjs.com/docs/api/promise.map.html